index.js

```diff
@@ -38,73 +40,96 @@ function arg(opts, {argv, permissive = false} = {}) {
-			while (argName in aliases) {
-				argName = aliases[argName];
-			}
+			for (let j = 0; j < separatedArguments.length; j++) {
+				const arg = separatedArguments[j];
+				const [originalArgName, argStr] = arg[1] === '-' ? arg.split('=', 2) : [arg, undefined];
 
-			if (!(argName in handlers)) {
-				if (permissive) {
-					result._.push(arg);
-					continue;
-				} else {
-					const err = new Error(`Unknown or unexpected option: ${originalArgName}`);
-					err.code = 'ARG_UNKNOWN_OPTION';
-					throw err;
+				let argName = originalArgName;
+				while (argName in aliases) {
+					argName = aliases[argName];
 				}
-			}
 
-			/* eslint-disable operator-linebreak */
-			const [type, isArray] = Array.isArray(handlers[argName])
-				? [handlers[argName][0], true]
-				: [handlers[argName], false];
-			/* eslint-enable operator-linebreak */
+				if (!(argName in handlers)) {
+					if (permissive) {
+						result._.push(arg);
+						continue;
+					} else {
+						const err = new Error(`Unknown or unexpected option: ${originalArgName}`);
+						err.code = 'ARG_UNKNOWN_OPTION';
+						throw err;
+					}
+				}
+
+				/* eslint-disable operator-linebreak */
+				const [type, isArray] = Array.isArray(handlers[argName])
+					? [handlers[argName][0], true]
+					: [handlers[argName], false];
+				/* eslint-enable operator-linebreak */
 
-			let value;
-			if (type === Boolean) {
-				value = true;
-			} else if (argStr === undefined) {
-				if (argv.length < i + 2 || (argv[i + 1].length > 1 && argv[i + 1][0] === '-')) {
-					const extended = originalArgName === argName ? '' : ` (alias for ${argName})`;
-					throw new Error(`Option requires argument: ${originalArgName}${extended}`);
+				if (!(type === Boolean || type[flagSymbol]) && ((j + 1) < separatedArguments.length)) {
+					throw new TypeError(`Option requires argument (but was followed by another short argument): ${originalArgName}`);
 				}
 
-				value = type(argv[i + 1], argName, result[argName]);
-				++i;
-			} else {
-				value = type(argStr, argName, result[argName]);
-			}
+				let value;
+				if (type === Boolean) {
+					value = true;
+				} else if (type[flagSymbol]) {
+					value = type(true, argName, result[argName]);
+				} else if (argStr === undefined) {
+					if (argv.length < i + 2 || (argv[i + 1].length > 1 && argv[i + 1][0] === '-')) {
+						const extended = originalArgName === argName ? '' : ` (alias for ${argName})`;
+						throw new Error(`Option requires argument: ${originalArgName}${extended}`);
+					}
+
+					value = type(argv[i + 1], argName, result[argName]);
+					++i;
+				} else {
+					value = type(argStr, argName, result[argName]);
+				}
 
-			if (isArray) {
-				if (result[argName]) {
-					result[argName].push(value);
+				if (isArray) {
+					if (result[argName]) {
+						result[argName].push(value);
+					} else {
+						result[argName] = [value];
+					}
 				} else {
-					result[argName] = [value];
+					result[argName] = value;
```

Comment:
To help decode what really happened here:

The condensation parsing code really happens on lines 57-59, which breaks up a `-abcd` argument into `[-a, -b, -c, -d]` and then runs the argument parsing step for each of those. Aside from a few extra checks and changing some variable names, that's really all that changed. The big diff is mostly due to shifting in the bulk of the logic by one indentation.